### ISSUE TITLE
Segfaults

### ISSUE THREAD

**Issue by eli-osherovich** [catboost/catboost#1935]
Problem: Catboost segfaults during training
catboost version:1.0.3
Operating System:Debian
CPU:Intel I7
GPU:

**Comment by eli-osherovich** [comment]
Catboost reliably segfaults with a certain set of parameters.

Sometimes it's a silent segfault, sometimes it is #1436 (a hairy error message). It depends on the random_state i set in the beginning....

**Comment by eli-osherovich** [comment]
My very wild guess is that the segfaults are related to (low values of) the  l2 regularization weight. When I remove it from the optuna search space I do not see it, when I raise the lower bound from 1e-8 to something like 1e-5 I do not see it either....

Of course, it is not a firm proof, but still....

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Hi @eli-osherovich !
Could you please check whether dev_efb_max_buckets=0 fixes the segfault?

**Comment by eli-osherovich** [comment]
Hi @Evgueni-Petrov-aka-espetrov , unfortunately it does not help.
Here is the set of parameters that segfaults:

```json
{
  "task_type": "CPU",
  "thread_count": 4,
  "random_state": 42,
  "learning_rate": 0.025,
  "l2_leaf_reg": 2.444914148737743e-06,
  "bootstrap_type": "Bernoulli",
  "grow_policy": "SymmetricTree",
  "max_depth": 5,
  "min_data_in_leaf": 1,
  "one_hot_max_size": 10,
  "fold_permutation_block": 1,
  "boosting_type": "Plain",
  "od_type": "Iter",
  "max_bin": 32,
  "feature_border_type": "GreedyLogSum",
  "nan_mode": "Min",
  "verbose": True,
  "allow_writing_files": False,
  "dev_efb_max_buckets": 0,
  "num_trees": 200,
  "objective": "Logloss",
  "eval_metric": "AUC",
  "od_wait": 100
}
```

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@eli-osherovich could you please provide a reproducer?

**Comment by eli-osherovich** [comment]
@Evgueni-Petrov-aka-espetrov sorry, cannot release company's data. 
Do you use fuzzy testing for catboost?

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
of course we do :)
any stats that you could kindly share so that we could try to generate a synthetic dataset similar to yours?
e.g. number of float/cat features, number of samples, number of different values per cat features?
how the dataset is passed to CatBoost -- numpy, pandas, tab-delimited file, previously saved CatBoost.Pool?
any similar info would be of use :)

**Comment by eli-osherovich** [comment]
@Evgueni-Petrov-aka-espetrov 

That would require some non-negligible amount of work on my side.... Will do it only if absolutely necessary. 
Meanwhile I downgraded catboost to the version `1.0.2` and my segfaults are gone! Since it's a random process, this does not necessary mean that the bug was introduced in `1.0.3`, but still worth checking.

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
the CatBoost logic in 1.0.2 and 1.0.3 is the same.
thus, highly likely, 1.0.2. has the same bug.
unless we catch the bug within 1 day of training on small random datasets, with the parameters that you indicate above, we will close this issue.

**Comment by eli-osherovich** [comment]
Unfortunately this is true....
Increased the number of trials in optuna and got a failure again....


```
FAIL (2021-11-29T20:59:57.073412+0200): This should be unreachable
  catboost/private/libs/algo/tensor_search_helpers.cpp:99
  GetSplit() failed
??+0 (0x7F2618641709)
??+0 (0x7F261863F718)
??+0 (0x7F2618CA29BB)
??+0 (0x7F2618CA2622)
??+0 (0x7F2618D6D26A)
??+0 (0x7F2618D2C08E)
??+0 (0x7F26186C2979)
??+0 (0x7F26186C60AE)
??+0 (0x7F26186C5D4E)
??+0 (0x7F26186C5A9E)
??+0 (0x7F26186AACB6)
??+0 (0x7F26186A7B4A)
??+0 (0x7F26186B281B)
??+0 (0x7F26186B419D)
??+0 (0x7F26186B40F6)
??+0 (0x7F2647AADD7E)
clone+63 (0x7F264787CCAF)
```

**Comment by eli-osherovich** [comment]
After trying the recent versions `1.0.x`, which all had segmentation faults I downgraded catboost to `0.26.1` and so far had no segfaults.

Please have a look at the difference.

**Comment by ThomasWolf0701** [comment]
Same problem here. Also see the segfaults only with 1.0.x and not before. Also use optuna to tune the catboost model and the segfaults occur more or less randomly. Would be great if there would be a fix, as before 1.0.x everything was working fine.

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@eli-osherovich @ThomasWolf0701 
We were running the following script for 24h + 24h (samples in 100 .. 200 + samples in 1000 .. 10000), and did not see any segmentation faults.
Would be grateful for a reproducer or some hints on the datatset.

```python
import catboost
import numpy

numpy.random.seed(20211130)
count = 0

while True:
    print(str(count) + '\r', end='')
    model = catboost.CatBoostClassifier(
        task_type = "CPU",
        thread_count = 4,
        random_state = 42,
        learning_rate = 0.025,
        l2_leaf_reg = 2.444914148737743e-06,
        bootstrap_type = "Bernoulli",
        grow_policy = "SymmetricTree",
        max_depth = 5,
        min_data_in_leaf = 1,
        one_hot_max_size = 10,
        fold_permutation_block = 1,
        boosting_type = "Plain",
        od_type = "Iter",
        max_bin = 32,
        feature_border_type = "GreedyLogSum",
        nan_mode = "Min",
        verbose = True,
        allow_writing_files = False,
        dev_efb_max_buckets = 0,
        num_trees = 200,
        objective = "Logloss",
        eval_metric = "AUC",
        od_wait = 100
    )

    factors = numpy.random.randint(1, 20)
    samples = numpy.random.randint(1000, 10000)
    cat = numpy.random.randint(0, factors, 5)
    train = numpy.random.randint(0, 10, (samples, factors))  # [0, 10)
    label = numpy.random.randint(0, 2, samples)  # [0, 2)
    try:
        model.fit(
            X=train, y=label,
            cat_features=cat,
            use_best_model=False,
            verbose=False
        )
    except Exception as e:
        print('Failed at step {} with {}'.format(count, e))
    count += 1
```

**Comment by eli-osherovich** [comment]
I will check if I can do something. The difficulty lies in the complex flow that predates training. The data undergoes a bunch of transformations before it enters catboost. Namely, what catboost gets is not stored on a disk.....

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
BTW, it may make sense to plug into your process CatBoost shared object built with --sanitize=memory or address... (if time permits, of course).

**Comment by LiuZhenkun-DUFE** [comment]
> 
Hello, have you adopted a custom loss function? I made the same mistake, and I'm not sure if the custom loss function is causing the problem or the CatBoost itself.

**Comment by alexwwang** [comment]
I also met the segmentation fault these days.
I use the monotonic1 dataset and it happens each time I run the program.
The code is attached. You may have a try.

[cb_test_modeling.py.txt](https://github.com/catboost/catboost/files/8183627/cb_test_modeling.py.txt)

Env info
catboost: v1.0.4
optuna: v2.10.0
python: v3.8.8
OS: macosx 10.15.7 / linux 3.10.0-1160.15.2.el7.x86_64

**Comment by datason** [comment]
I have the same error:
**It depends on random_seed of CatBoostRegressor** and data distribution, ofc

`FAIL (2022-06-14T11:31:48.095363+0300): This should be unreachable
  C:\Program Files (x86)\Go Agent\pipelines\BuildMaster\catboost.git\catboost\private\libs\algo\tensor_search_helpers.cpp:99
  TCandidateInfo::GetSplit() failed
PyInit__catboost+1852936 (0x7FFEECB08718)
PyInit__catboost+1948791 (0x7FFEECB1FD87)
PyInit__catboost+1916955 (0x7FFEECB1812B)
PyInit__catboost+6420388 (0x7FFEECF638B4)
PyInit__catboost+6419308 (0x7FFEECF6347C)
PyInit__catboost+6546394 (0x7FFEECF824EA)
PyInit__catboost+6325357 (0x7FFEECF4C57D)
PyInit__catboost+2491980 (0x7FFEECBA475C)
PyInit__catboost+2495344 (0x7FFEECBA5480)
PyInit__catboost+2499423 (0x7FFEECBA646F)
PyInit__catboost+2408715 (0x7FFEECB9021B)
PyInit__catboost+2402044 (0x7FFEECB8E80C)
PyInit__catboost+2436962 (0x7FFEECB97072)
PyInit__catboost+2474495 (0x7FFEECBA030F)
PyInit__catboost+2475419 (0x7FFEECBA06AB)
PyInit__catboost+13330740 (0x7FFEED5FAA44)
BaseThreadInitThunk+20 (0x7FFFD6507034)
RtlUserThreadStart+33 (0x7FFFD8022651)

Process finished with exit code -1073740791 (0xC0000409)`

**Comment by sktin** [comment]
A recent post in a kaggle competition reports a similar problem. I extracted just enough code from the OP's notebook to demonstrate a crash. It may help to provide a concrete repro to finally get this bug fixed. The repro uses kaggle datasets.

Link to the post: https://www.kaggle.com/competitions/playground-series-s4e10/discussion/542912#3029934

Datasets:
1. https://www.kaggle.com/competitions/playground-series-s4e10/data
2. https://www.kaggle.com/datasets/chilledwanker/loan-approval-prediction/data
